jax/_src/api.py

```diff
@@ -3245,6 +3246,50 @@ def try_to_block(x):
       return x
   return jax.tree_util.tree_map(try_to_block, x)
 
+def pure_callback(callback: Callable[..., Any], result_shape_dtypes: Any,
+                  *args: Any, **kwargs: Any):
+  """Calls a pure Python callback function from staged out JAX programs.
+
+  ``pure_callback`` enables calling a Python function in JIT-ed JAX functions.
+  The input ``callback`` will be passed NumPy arrays in place of JAX arrays and
+  should also return NumPy arrays. Execution takes place on the CPU host.
+
+  The callback is treated as "pure" meaning it can be called multiple times when
+  transformed (for example in a ``vmap`` or ``pmap``), and it can also
+  potentially be removed from JAX programs via dead-code elimination. Pure
+  callbacks can also be reordered if data-dependence allows.
+
+  When ``pmap``-ed, the pure callback will be called several times (one on each axis
+  of the map). When `vmap`-ed the behavior will depend on the value of the
+  ``rank_polymorphic`` keyword argument. If the callback is indicated as rank
+  polymorphic, the callback will be called directly on batched inputs (where the
```

Comment:
There could be more than one batch axis (with multiple applications of `vmap`), but the text here talks about _the_ batch axis. Maybe tweak to highlight that there could be arbitrarily many batch axes?

I wonder if we can make the name of the option more descriptive, like instead of `rank_polymorphic` we can call it `arbitrary_leading_batch_axes` or something. Ideas?

Reply:
Yea I think rank-polymorphic is not the best term here. Another option is maybe "vectorized" (borrowing from `np.vectorize`)?